### ISSUE TITLE
Strange behavior/animation when adding/Removing items

### ISSUE THREAD

**Issue by tamimattafi** [marcorei/Infinite-Fire#9]
i am trying to create a chat `activity` so that users can send messages to each other.. i'm using `Firebase Database` in this case, with [marcorei/Infinite-Fire][1] as an `Adapter`.. The **problem** is when ever the user sends/receives a message, the `RecyclerView` starts blinking and doing some other strange behaviors.. but that only happens when there are only few items.. after that everything gets perfect.
well , i'm also using `FirebaseDatabase.getInstance().setPersistenceEnabled(true);`
so the user can access his data even if he's offline. i think that the problem is when the user clicks send.. the message is pending/stored in the app cash and in few moments it will be stored to the Database , when this happens.. the `RecyclerView` refresh the last item.

> **Note** : I Uploaded a short video to YouTube , so it can describe the issue more efficiently. [Click Here To Watch The Video][2] an other type is shown here [Click Here](https://youtu.be/Q9qg-ZSM268)

i tried a lot of codes but none have worked/was enough.. so here's what i did,

1- **Setting item animator to null** : 

    mRecyclerView.setItemAnimator(null);

2- **Setting Supports Change animation to false** : 

    RecyclerView.ItemAnimator animator = mRecyclerView.getItemAnimator();
        if (animator instanceof SimpleItemAnimator) {
            ((SimpleItemAnimator) animator).setSupportsChangeAnimations(false);
        }

3- **Setting Add/Remove duration to 0** :

    mRecyclerView.getItemAnimator().setAddDuration(0);
    mRecyclerView.getItemAnimator().setRemoveDuration(0);

4- **Setting stable IDs for the adapter to true** :
but that prevents the adapter to show more items , it only updates the last item to the newest added one.

    adapter.setHasStableIds(true);

> Here is my code : 

 - **adapter onBindViewHolder** : 
 

        public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {

            int viewType = getItemViewType(position);

            switch(viewType) {

                case VIEW_TYPE_CONTENT:

                    final RecyclerViewAdapter.posts_holder contentHolder = (RecyclerViewAdapter.posts_holder) holder;
                    message message = snapshots.getItem(position - indexOffset).getValue();
                    String userUid = FirebaseAuth.getInstance().getCurrentUser().getUid();
                    if (message.getMessageSender().equals(userUid)){
                        contentHolder.mSender.setText(message.getMessageText());
                        contentHolder.mRecieverRoot.setVisibility(View.GONE);
                        return;
                    }
                    if (!message.getMessageSender().equals(userUid)){
                        contentHolder.mReciever.setText(message.getMessageText());
                        contentHolder.mSenderRoot.setVisibility(View.GONE);
                        return;
                    }
                    break;

                case VIEW_TYPE_HEADER:

                    RecyclerViewAdapter.LoadingHolder headerHolder = (RecyclerViewAdapter.LoadingHolder) holder;

                    headerHolder.progressBar.setVisibility((isInitiallyLoading()) ? View.VISIBLE : View.GONE);

                    break;

                case VIEW_TYPE_FOOTER:

                    RecyclerViewAdapter.LoadingHolder footerHolder = (RecyclerViewAdapter.LoadingHolder) holder;

                    footerHolder.progressBar.setVisibility((isLoadingMore()) ? View.VISIBLE : View.GONE);

                    break;

                default:

                    throw new IllegalArgumentException("Unknown type");

            }

        }

 - **Database Reference , Query and MessagesArray** :

        mDatabaseRefrance = FirebaseDatabase.getInstance().getReference();
        mQuery = mDatabaseRefrance.child("chat").child(mUser.getUid()).child   (getUserUid()).orderByKey();
        mMessagesArray = new InfiniteFireArray<>(
                message.class,
                mQuery,
                50,
                25,
                false,
                false

        );
        adapter = new RecyclerViewAdapter(mMessagesArray);

 - **Adapter Data Observer** :

        adapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onItemRangeInserted(int positionStart, int itemCount) {
                int position = manager.findFirstVisibleItemPosition();
                if (position > 3){
                    Toast.makeText(Chat.this , " You Have A New Message " , Toast.LENGTH_LONG).show();
                    return;
                }
                else {
                    manager.smoothScrollToPosition(mRecyclerView,null,1);
                }

            }
        });

 - **Database Structure** : 

 [![enter image description here][3]][3]

  [1]: https://github.com/marcorei/Infinite-Fire
  [2]: https://www.youtube.com/watch?v=V433jQYRDpg
  [3]: https://i.stack.imgur.com/jurAN.png

**Comment by marcorei** [comment]
This looks strange. I guess this is caused by events fired by `Firebase`. `messageTime` looks like it's a server-generated timestamp which would cause `Firebase` to update the message soon after you created it.

I'd suggest copying the [InfiniteFireRecyclerViewAdapter](https://github.com/marcorei/Infinite-Fire/blob/master/infinitefire/src/main/java/com/marcorei/infinitefire/InfiniteFireRecyclerViewAdapter.java) and logging the [onChanged](https://github.com/marcorei/Infinite-Fire/blob/master/infinitefire/src/main/java/com/marcorei/infinitefire/InfiniteFireRecyclerViewAdapter.java#L50) events to see if they correlate with the blinking.

**Comment by tamimattafi** [comment]
@marcorei That's correct,  messageTime is a server-generated timestamp and I've mentioned that firebase might update the message soon after i create it 

> " i think that the problem is when the user clicks send.. the message is pending/stored in the app cash and in few moments it will be stored to the Database , when this happens.. the RecyclerView refresh the last item. "

i need the timestamp to show the users message time depending on each one's time zone.. using system time might cause a serious problem with that.
don't you think i should disable 
 ` case Changed:
                        notifyItemChanged(index + indexOffset);
                        break;`  ??
so the **RecyclerView** won't update that specific message. I'm also curious about why this happen only to the first 10-15 items then it becomes normal?

**Comment by marcorei** [comment]
No idea why the behavior changes after 10-15 items. The events fired by Firebase change after the total number of items exceed the limit of the query, but your initial page size is 50 so that shouldn't be an issue.

Based on your videos and the fact that the problem still exists even if you remove the `ItemAnimator` my guess would be that Firebase removes the new item and adds it again.

Can you reproduce the issue in a simple setup, for example with the example app in this repository?

(I agree that you should always use the server generated timestamp)
